safe_api.js

```diff
@@ -4,233 +4,295 @@ window.safeAPI = {};
-				};
-				// if both are defined and different, unclear how to proceed.
-				if ((mdHandle != undefined) && (this.mdHandle != undefined) && (mdHandle != this.mdHandle)) throw 'Cannot decide which handle to free!';
-			});
+		asNFS() {
+			return new NfsEmulation(_ => safeMutableData.emulateAs(this.handle, 'NFS'), this.root);
 		}
 
 	};
 
-	var Permissions = class extends HandleSynchronisedSafeObject {
+	/* not reviewed -BTB
+	var Permissions = class extends HandleSynchronisedSafeChild {
 
 		insertPermissionsSet(signKey, permissionsSet) {
 
-			this.handlePromise = Promise.all([this.handlePromise, signKey ? signKey.handlePromise : null, permissionsSet.handlePromise])
-				.then(([permHandle, skHandle, pSetHandle]) => {
-
-					console.log('skHandle:'+skHandle);
-					return safeMutableDataPermissions.insertPermissionsSet(permHandle, skHandle, pSetHandle)
-						.then(() => permHandle);
-				});
+			return this.all([this.handlePromise, signKey ? signKey.handlePromise : null, permissionsSet.handlePromise])
+			.then(([permHandle, skHandle, pSetHandle]) => {
+					console.log('skHandle: ',skHandle);
+					return safeMutableDataPermissions.insertPermissionsSet(permHandle, skHandle, pSetHandle);
+			})
+			.then(_ => permHandle);
 		}
 
 	};
+	*/
 
-	var Entries = class extends HandleSynchronisedSafeObject {
+	var Entries = class extends HandleSynchronisedSafeChild {
+
+		len() {
+			return this.then(_ => safeMutableDataEntries.len(this.handle));
+		}
 
 		insert(key, value) {
-			this.handlePromise = this.handlePromise.then((entriesHandle) => safeMutableDataEntries.insert(entriesHandle, key, value)
-				.then(() => entriesHandle));
+			return this.then(_ => safeMutableDataEntries.insert(this.handle, key, value));
+		}
+
+		forEach(f) {
+			return this.then(_ => safeMutableDataEntries.forEach(this.handle, f));
 		}
 
 	};
 
-	var PermissionsSet = class extends HandleSynchronisedSafeObject {
+	var PermissionsSet = class extends HandleSynchronisedSafeChild {
 
 		setAllow(action) {
-			this.handlePromise = this.handlePromise.then((pSetHandle) => safeMutableDataPermissionsSet.setAllow(pSetHandle, action)
-				.then(() => pSetHandle));
+			return this.then((pSetHandle) => safeMutableDataPermissionsSet.setAllow(pSetHandle, action))
+			.then(_ => pSetHandle);
 		}
 
 	};
 
-	var Mutation = class extends HandleSynchronisedSafeObject {
+	var Mutation = class extends HandleSynchronisedSafeChild {
 
 		insert(key, value) {
-			this.handlePromise = this.handlePromise.then((mutHandle) => safeMutableDataMutation.insert(mutHandle, key, value)
-				.then(() => mutHandle));
+			return this.then((mutHandle) => safeMutableDataMutation.insert(mutHandle, key, value))
+			.then(_ => mutHandle);
 		}
 
 	};
 
-	var SignKey = class extends HandleSynchronisedSafeObject {
+	var NfsEmulation = class extends HandleSynchronisedSafeChild {
+
+		create(content) {
+			return new NfsFile(_ => safeNfs.create(this.handle, content), this);
```

Comment:
shouldn't it be `this.root` instead of `this` as the last parameter?

Reply:
Yes, it should. Good catch!

Reply:
At some point, I plan to write some bit of code that automates `new X(_ => Y(this.handle, Z), this.root` so that creating new objects would be more like `this.new(X, Y, Z)`. The biggest issue is handling Z. I know there is a way to handle argument lists in JavaScript, I just need to look up how. I think `...` is part of the solution.

Reply:
Ugh, this typo actually explains a lot of bad behavior in my test code. There were no errors, but I think it causes an infinite loop. That's why some calls seemed to vanish, rather than logging something or "undefined" to console. I also recall my Browser going a bit slower. This is a pretty serious mistake that I missed!